### PR TITLE
[flang] Add notify-type and notify-wait-stmt

### PR DESCRIPTION
**Pull request by ktras**
Add `notify-type` to `iso_fortran_env` module. Add `notify-wait-stmt` to the parser and add checks for constraints on the statement, `C1177` and `C1178`, from the Fortran 2023 standard. Add three semantics tests for `notify-wait-stmt`.

### CODE REVIEW

File: flang/include/flang/Parser/parse-tree.h [hunk 1]
```diff
@@ -2492,6 +2494,16 @@ struct StopStmt {
   std::tuple<Kind, std::optional<StopCode>, std::optional<ScalarLogicalExpr>> t;
 };
 
+// F2023: R1167 notify-variable -> scalar-variable
+using NotifyVariable = Scalar<Variable>;
```
Comment (klausler) [inline flang/include/flang/Parser/parse-tree.h:2498]: Just use `Scalar<Variable>` in the tuple below; the new type here doesn't add value.

File: flang/lib/Semantics/check-coarray.cpp [hunk 2]
```diff
@@ -212,32 +195,76 @@ void CoarrayChecker::Leave(const parser::EventWaitStmt &x) {
...
 }
 
+void CoarrayChecker::Leave(const parser::NotifyWaitStmt &x) {
+  const auto &notifyVar{std::get<parser::NotifyVariable>(x.t)};
+
+  if (const auto *expr{GetExpr(context_, notifyVar)}) {
+    if (ExtractCoarrayRef(expr)) {
+      context_.Say(parser::FindSourceLocation(notifyVar), // F2023 - C1178
+          "A notify-variable in a NOTIFY WAIT statement may not be a coindexed object"_err_en_US);
+    } else {
```
Comment (klausler) [inline flang/lib/Semantics/check-coarray.cpp:231]: `else if` would be better here

**Review by klausler — APPROVED** [review]
really great work here, thank you!

File: flang/include/flang/Parser/parse-tree.h (hunk 1 above)
Comment (ktras) [reply flang/include/flang/Parser/parse-tree.h]: @klausler Thanks for the review, comment addressed.

File: flang/lib/Semantics/check-coarray.cpp (hunk 2 above)
Comment (ktras) [reply flang/lib/Semantics/check-coarray.cpp]: @klausler Thanks for the review, comment addressed.

**Review by psteinfeld — APPROVED** [review]
All builds and tests correctly and looks good.

**Pull request merged**